### PR TITLE
#470 initial pipeline support

### PR DESCRIPTION
**Pull request by jeffastorey**
Changes to drone core to support the pipelines https://github.com/drone/drone/issues/470. Full support requires changes to drone-exec, which I'll committing shortly.

### CODE REVIEW

File: controller/build.go [hunk 1]
```diff
@@ -236,6 +236,8 @@ func PostBuild(c *gin.Context) {
 	build.Started = 0
 	build.Finished = 0
 	build.Enqueued = time.Now().UTC().Unix()
+	build.Event = c.DefaultQuery("event", build.Event)
+	build.Deploy = c.DefaultQuery("deploy_to", build.Deploy)	
 	for _, job := range jobs {
```
Comment (bradrydzewski) [inline controller/build.go]: looks good!

we should probably just make sure that `build.Event` is equal to one of these values
https://github.com/drone/drone/blob/master/model/const.go#L3

Comment (bradrydzewski) [inline controller/build.go]: we should probably also only enable overriding `build.Event` and `build.Deploy` if `fork` is true. The reason being when `fork` is false it just re-runs the existing build number and does not create a new build number. We probably want to treat the event end deploy values as immutable unless being forked.

**Comment by bradrydzewski** [comment]
Looks great. I added a few minor notes for validation. I'm sorry I didn't bring them up previously, however, they hadn't crossed my mind until I started the actual code review.

**Comment by jeffastorey** [comment]
Those make sense. In what situations is a build actually forked? Does the restart build button do it or only through API calls?

Related to this, is it possible to access/use the information that's passed as part of the deploy hook? For example I just wanted a single drone deployment configuration whose branch is based on the name passed in the deployment event, is that possible?

**Comment by bradrydzewski** [comment]
> Those make sense. In what situations is a build actually forked? Does the restart build button do it or only through API calls?

The restart button restarts the build and does not fork (assign new build number). The api and command line utility can be used to fork and re-execute the build:

```
drone build fork octocat/hello-world 55
```

The primary reason for this behavior is the restart button in the user interface is intended for quick restarts due to some external failure (ie npm was down while executing the build, so I need to re-run). The fork functionality was created to enable nighty builds from cron jobs, using the cli, where a new build number is desirable.

> Related to this, is it possible to access/use the information that's passed as part of the deploy hook? For example I just wanted a single drone deployment configuration whose branch is based on the name passed in the deployment event, is that possible?

not sure I'm 100% clear. Can you provide some examples or mockup yaml? Which data and how you'd like to use?

**Comment by jeffastorey** [comment]
Sample YAML might be something like this. The idea being that I can use a bucket name that is the same as the environment in this case, rather than creating separate deploy configurations for each of my environments.

```
deploy:
    terraform:
        plan: false
        remote:
            backend: S3
            config:
                bucket: <environment-from-deploy>
                key: tfstate
        when:
            event: deployment
            environment: mytestenvironment
```

**Comment by jeffastorey** [comment]
Also I'm going to close this PR and reopen where I'm not trying to merge in my master branch to avoid the situation where I tried to open 2 PRs on master.

**Pull request closed without merging**